### ISSUE TITLE
Implement Parser

### ISSUE THREAD

**Issue by Pyronix** [Pyronix/imp-go#5]
https://sulzmann.github.io/ModelBasedSW/extended-parser.html#(6)

**Comment by Pyronix** [comment]
Official syntax (variant with `prog ::= statement`):
```
vars       Variable names, start with lower-case letter

prog      ::= statement
block     ::= "{" statement "}"
statement ::=  statement ";" statement           -- Command sequence
            |  vars ":=" exp                     -- Variable declaration
            |  vars "=" exp                      -- Variable assignment
            |  "while" exp block                 -- While
            |  "if" exp block "else" block       -- If-then-else
            |  "print" exp                       -- Print

exp ::= 0 | 1 | -1 | ...     -- Integers
     | "true" | "false"      -- Booleans
     | exp "+" exp           -- Addition
     | exp "*" exp           -- Multiplication
     | exp "||" exp          -- Disjunction
     | exp "&&" exp          -- Conjunction
     | "!" exp               -- Negation
     | exp "==" exp          -- Equality test
     | exp "<" exp           -- Lesser test
     | "(" exp ")"           -- Grouping of expressions
     | vars                  -- Variables
```

**Comment by Pyronix** [comment]
Needs to be transformed to eliminate left recursion and resolve precedence.
Will be done accordingly to this book:
http://craftinginterpreters.com/parsing-expressions.html

Let's focus on `exp` first.

First we define the precedence rules as following, going from lowest to highest:
- Disjunction: exp "||" exp
- Conjunction: exp "&&" exp
- Equality: exp "==" exp
- Comparison: exp "<" exp
- Term: exp "+" exp
- Factor: exp "*" exp
- Unary: "!" exp

We use the rules to derive a new grammar.
Each rule only matches expressions at its precedence level or higher.
```
expression ::= disjunction
disjunction ::= conjunction ("||" conjunction)*
conjunction ::= equality ("&&" equality)*
equality ::= comparison ("==" term)*
comparison ::= term ("<" term)*
term ::= factor ("+" factor)*
factor ::= unary ("*" unary)*
unary ::= "!" unary | primary
primary ::= INT | BOOL | "(" expression ")" | vars 
```

Rules for statements only need minimal modification, we just need to eliminate the left recursion for a command sequence:
```
prog ::= sequence
block     ::= "{" statement "}"
sequence ::= statement (";" statement)*
statement ::= vars ":=" expression | ...
```

**Comment by Pyronix** [comment]
Closed with #7
